ath79: add support for WD My Net N750

rmounce opened the pull request.

ynezz commented on target/linux/ath79/dts/ar9344_wd_mynet-n750.dts:
```diff
@@ -0,0 +1,221 @@
+	phy-mask = <0>;
+
+	/* PHY 0, MII_BMCR, (BMCR_RESET | BMCR_ANENABLE | BMCR_SPEED1000) */
+	qca,ag71xx-mdio-resetvals = <0 0x00 0x9040>;
```
This doesn't look right to me as you're using ar8xxx regs (BMCR) in the ar9344 MDIO block (AG71XX). Probably ar8xxx_phy_check_aneg needs some changes in order to support your device properly as well?

rmounce replied:
See 8b712436f37cea6b25e5a717042ab10b85fab3be and c72313fd9801143029e7381d3df92e49fe45c56b

This register needs to be written to allow the switch to be discovered, before the appropriate ar8xxx driver is in play. Otherwise there is a chicken and egg problem.

Only phy 0 needs to be brought up in my testing, not 0-5 per the previous WD N750 fixup.

ynezz replied:
FYI @olerem has almost [upstreamed the ag71xx driver](https://patchwork.ozlabs.org/patch/1101843/) so I would like to be really careful now and accept only changes to ag71xx which doesn't make his effort harder. I've asked him if he could chime in and provide you with some feedback as well.

From my point of view, kernel folks won't accept this solution as you're fiddling with ar8xxx device inside the ag71xx. Simply ag71xx don't need to care what the underlying ar8xxx switch needs, it should be handled by ar8xxx switch driver itself.

It looks like the call stack is following:

```
phy_probe
  phy_device_reset
     mdio_device_reset
        ar8xxx_phy_soft_reset
           ar8xxx_phy_config_init
              ar8xxx_phy_check_aneg
  ar8xxx_phy_probe
```

So I'm either reading it wrong or I see several places where this could be fixed in ar8xxx.

olerem replied:
I agree, the code has nothing to do with ag71xx. It is poorly ar8xxx code and should be implemented there.

981213 replied:
If this problem is just that PHY ID isn't readable before some register operations, you could take a look at my approach for TP-LINK TL-WR941N v7 CN where I added a mdio-device probing for ar8216 switch driver.

chunkeey replied:
You can force the phy-id discovery see:

https://www.kernel.org/doc/Documentation/devicetree/bindings/net/phy.txt

```
  If the PHY reports an incorrect ID (or none at all) then the
  "compatible" list may contain an entry with the correct PHY ID in the
  form: "ethernet-phy-idAAAA.BBBB" where
     AAAA - The value of the 16 bit Phy Identifier 1 register as
            4 hex digits. This is the chip vendor OUI bits 3:18
     BBBB - The value of the 16 bit Phy Identifier 2 register as
            4 hex digits. This is the chip vendor OUI bits 19:24,
            followed by 10 bits of a vendor specific ID.
```

rmounce replied:
phy-id did the trick. Thank you all for pointing me in the right direction.

ynezz replied:
I find it quite entertaining, that we've currently (at least?) two ways how to approach this issue. It makes me wonder @981213 if we need both (upstream phy based one and yours mdio one) probably very similar solutions.

I mean, is upstream going to accept the mdio based one if probably the same could be achieved with phy based one, which is already available in the upstream? Maybe it would make sense to convert TL-WR941N v7 CN to that phy based upstream solution and remove the mdio based probing?

981213 replied:
@ynezz 
Technically a switch isn't a PHY. mdio-device is the upstream framework specifically for these devices that uses mdio bus but isn't a PHY, and that's what's used for almost all upstream DSA drivers. (The only exception is the one for the ancient Marvell 88E6060)

Also, the added mdio-device probing isn't only for WR941N v7. The builtin switch in ar724x/ar933x uses the same PHY ID and chip ID as standalone AR8216 switch while they are actually different, and we need a way to distinguish them (where the compatible string is used for this purpose).

ynezz commented on target/linux/ath79/dts/ar9344_wd_mynet-n750.dts:
```diff
@@ -0,0 +1,221 @@
+		};
+	};
+
+	gpio-export {
```
We're trying to get rid of the `gpio-export` functionality as it's hack for missing kernel functionality, which was rejected in upstream kernel long time ago, for details see this email http://lists.infradead.org/pipermail/openwrt-devel/2019-February/015772.html, discussion in #1366 or https://github.com/openwrt/openwrt/pull/1814#issuecomment-462942022.

So I'm wondering if it would be possible to replace the `gpio-export` construct with the userspace gpio-switch alternative, for example https://github.com/openwrt/openwrt/blob/master/target/linux/layerscape/base-files/etc/board.d/03_gpio_switches#L20-L22

ynezz commented on target/linux/ath79/dts/ar9344_wd_mynet-n750.dts:
```diff
@@ -0,0 +1,221 @@
+		gpio_ext_lna0 {
+			gpio-export,name = "wd:ext:lna0";
+			gpio-export,output = <1>;
+			gpios = <&gpio 15 GPIO_ACTIVE_HIGH>;
```
Maybe for this it would be better to use gpio-hog instead?

rmounce replied:
Changed to gpio-hog and tested, works fine.
TP-Link WDR4300 should probably be updated to match.

ynezz commented on target/linux/ath79/dts/ar9344_wd_mynet-n750.dts:
```diff
@@ -0,0 +1,221 @@
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
```
can you please try to change this to `gpio-keys` (interrupt based driver)

rmounce replied:
Changed and tested, works fine.

ynezz commented on target/linux/ath79/image/generic.mk:
```diff
@@ -655,6 +655,17 @@ define Device/rosinson_wr818
+  $(Device/seama)
+  ATH_SOC := ar9344
+  DEVICE_TITLE := Western Digital My Net N750
+  IMAGE_SIZE := 15808k
```
0xf80000 = 15872k ?

rmounce replied:
I was following the ar71xx profile here - didn't want to take a chance for the sake of 0.4% extra space.

I've tested with 15872k and it works fine, inc. sysupgrade from ar71xx preserving config.

ynezz replied:
Well, ar71xx could be wrong, I've just checked IMAGE_SIZE as it should correspond with the firmware partition size otherwise it simply doesn't make sense and might be confusing and should be probably noted in the commit message.

981213 commented on target/linux/ath79/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom:
```diff
@@ -184,6 +182,10 @@ case "$FIRMWARE" in
 		;;
+	wd,mynet-n750)
+		ath9k_eeprom_extract "art" 20480 1088
+		ath9k_patch_fw_mac_crc $(mtd_get_mac_ascii devdata "wlan5mac") 2
```
There is no CRC field in AR9380 eeprom, ath9k_patch_fw_mac is enough here.

981213 commented on target/linux/ath79/dts/ar9344_wd_mynet-n750.dts:
```diff
@@ -0,0 +1,216 @@
+		compatible = "gpio-leds";
+
+		wifi {
+			label = "wd:blue:wifi";
```
LED naming scheme is {boardname}:{color}:{name} so it should be prefixed with mynet-n750 here.
TP-Link is an exception here because they have many boards sharing the same LED configuration.

981213 commented on target/linux/ath79/dts/ar9344_wd_mynet-n750.dts:
```diff
@@ -0,0 +1,216 @@
+&pcie {
+	status = "okay";
+
+	ath9k: wifi@0,0 {
```
This node isn't referenced elsewhere so the ath9k label can be dropped.

981213 reviewed (commented):
As said in previous comment, you can make use of the mdio-device probing added in 7d504f68a651f328ae78beccd2f647d6f62fe39b to replace the hack for ar8327 switch.

Some other nitpickings below :)

rmounce committed:
ath79: add support for WD My Net N750

SoC: AR9344
RAM: 128MB
Flash: 16MiB Winbond 25Q128BVFG SPI NOR
5GHz WiFi: AR9380 PCIe 3x3:3 802.11n
2.4GHz WiFi: AR9344 (SoC) AHB 2x2:2 802.11n

5x Gigabit ethernet via AR8327N switch (green + amber LEDs)
2x USB 2.0 via GL850G hub
4x front LEDs from SoC GPIO
1x front WPS button from SoC GPIO
1x bottom reset button from SoC GPIO

Known issues:
AR8327N LEDs only have default functionality, not presented in sysfs.
This is a regression from ar71xx.

UART header JP1, 115200 no parity 1 stop
TX
GND
VCC
(N/P)
RX

See https://openwrt.org/toh/wd/n750 for flashing detail.
Procedures unchanged from ar71xx.

Tested sysupgrade + factory flash from WD Emergency Recovery

Signed-off-by: Ryan Mounce <ryan@mounce.com.au>

981213 commented:
You should use the mdio-device approach instead. It doesn't rely on PHY ID for actual switch detection and you don't need that hack at all.

981213 commented on target/linux/ath79/dts/ar9344_wd_mynet-n750.dts:
```diff
@@ -0,0 +1,216 @@
+	pll-data = <0x06000000 0x00000101 0x00001616>;
+
+	phy-mode = "rgmii";
+	phy-handle = <&phy0>;
```
Replace this phy-handle line with the following fixed link:
```
	fixed-link {
		speed = <1000>;
		full-duplex;
	};
```

981213 commented on target/linux/ath79/dts/ar9344_wd_mynet-n750.dts:
```diff
@@ -0,0 +1,216 @@
+		compatible = "ethernet-phy-id004d.d033";
+		phy-mode = "rgmii";
+
+		qca,ar8xxx-phy-probe-reset;
```
Drop this line and your hack.

981213 commented on target/linux/ath79/dts/ar9344_wd_mynet-n750.dts:
```diff
@@ -0,0 +1,216 @@
+
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+		compatible = "ethernet-phy-id004d.d033";
```
Replace the above 3 lines with the following:
```
	switch0@1f {
		compatible = "qca,ar8327";
		reg = <0x1f>;
```

rmounce replied:
I'm curious
0x0 also seemed to work, what is the difference with 0x1f?

981213 replied:
There is no difference here.
It's supposed to be a PHY address used by the device, however there isn't one specific address used by the switch. 0x0-0x4 is used to access 5 switch PHYs and other addresses are used to access switch internal registers, so this reg field isn't used.

There is one address used for writing the upper part of the register address we want to access and I decided to use that one as PHY address for this node.
However I just noticed that it's 0x18 instead of 0x1f :(

981213 reviewed (commented):
To use the mdio-device probing, you need to do the following things:

rmounce commented:
> You should use the mdio-device approach instead. It doesn't rely on PHY ID for actual switch detection and you don't need that hack at all.

This works now! Thanks, it's a much cleaner approach.

I had tried this approach earlier and got switch communication working, however for whatever reason I assumed that fixed-link would not work and tried to mix and match the two approaches. ar8xxx_phy_config_init was unhappy.

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

adrianschmutzler commented:
Although this PR has been has been closed already, I would be interested in some additional data about the device(s) in case you still have access to it:

I'm currently working on a PR to store the label-mac-address in the DTS (https://github.com/openwrt/openwrt/pull/2159).

I would appreciate if you could provide information about your device:
Please run
`find /proc/device-tree/ -name "*mac-address*"`
on your device (just with the current firmware, no need to update anything; just has to be ath79, no ar71xx).
For each of the returned paths (if there are any), retrieve the mac-address with `cat`, e.g.
`cat /proc/device-tree/ahb/eth@19000000/mac-address`

Now check which of those returned MAC addresses matches the one on the label/sticker/case of your device and tell me the device-tree path leading to it. Note that you do not have to post your MAC addresses here, but just a statement about which match, e.g. like
```
/proc/device-tree/ahb/eth@1a000000/mac-address
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address-increment
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address
/proc/device-tree/ahb/eth@19000000/mac-address <== this one
/proc/device-tree/ahb/eth@19000000/mtd-mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mtd-mac-address
```

There may be one, two or no paths giving the correct address. It may also happen that the `find` command returns nothing. Thanks.

adrianschmutzler commented:
Unfortunately, my description is slightly incorrect. The MAC addresses are stored binary, so you can't read with `cat`, but have to use the following code to read from the paths retrieved by `find`:
```
. /lib/functions/system.sh
get_mac_binary "$yourpath" 0
```
The include is obviously only required once. Sorry.
